File: saul-examples/src/main/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/EntityMentionRelation/SL_SVM/iJLIS/Runner.scala
package edu.illinois.cs.cogcomp.saulexamples.nlp.EntityMentionRelation.SL_SVM.iJLIS

import java.util
import java.util.{ArrayList, List}

import edu.illinois.cs.cogcomp.saulexamples.EntityMentionRelation.reader.Conll04_ReaderNew
import edu.illinois.cs.cogcomp.sl.core._
import edu.illinois.cs.cogcomp.sl.learner.LearnerFactory
import edu.illinois.cs.cogcomp.sl.util.{Lexiconer, WeightVector}

object Runner {

  var lexm: Lexiconer = new Lexiconer()

  def readXY(cr:Conll04_ReaderNew, outlist: => List[IStructure], sclist: => List[IInstance], l:Int,u:Int){
    var i: Int = l
    while (i < u){//cr.instances.size){
    val newqi: ERqInstance = new ERqInstance(cr.instances.elementAt(i))
      sclist.add(newqi)
      val a: Nodevalues=new Nodevalues()
      a.a=(cr.instances.elementAt(i).entType.toLowerCase.contains("peop"))
      val cs: ERiStructure = new ERiStructure(newqi, a.a)
      outlist.add(cs)
      i=i+1
    }
  }

  def trainSSVM(modelname: String): Unit = {
    var cr = new Conll04_ReaderNew("./data/EntityMentionRelation/conll04.corp","Pair")
    var sclist: List[IInstance]= new ArrayList[IInstance]//=readX(cr)
    var outlist: List[IStructure] = new ArrayList[IStructure]//readY(cr)
    readXY(cr,outlist,sclist,1,100)

    cr=null

    lexm.setAllowNewFeatures(false)
    val sp: SLProblem = new SLProblem
    sp.instanceList = sclist
    sp.goldStructureList = outlist
    val para: SLParameters = new SLParameters
    para.TOTAL_NUMBER_FEATURE = lexm.getNumOfFeature
    para.C_FOR_STRUCTURE = 1
    para.CHECK_INFERENCE_OPT = false
    val model: SLModel = new SLModel
    model.featureGenerator= new ERFeatureGenerator
    model.lm = lexm

    model.config = new util.HashMap();
    model.para=para

    model.infSolver = new iERinference;
    val learner = LearnerFactory.getLearner(model.infSolver, model.featureGenerator, para);
    model.wv = learner.train(sp);
    model.saveModel(modelname);
  }

  def testSequenceSSVM(model_name: String) {
    val model: SLModel = SLModel.loadModel(model_name)
    Runner.lexm = model.lm
    val learned_wv: WeightVector = model.wv
    val inference_proc: iERinference = model.infSolver.asInstanceOf[iERinference]
    var cr = new Conll04_ReaderNew("./data/EntityMentionRelation/conll04.corp","Pair")
    var sclist: List[IInstance]= new ArrayList[IInstance]//=readX(cr)
    var outlist: List[IStructure] = new ArrayList[IStructure]//readY(cr)
    readXY(cr,outlist,sclist,101,200)
    cr=null

    val sp: SLProblem = new SLProblem
    sp.instanceList = sclist
    sp.goldStructureList = outlist
    printTestACC(sp, inference_proc, learned_wv)
  }
  private def printTestACC(sp: SLProblem, s_finder: AbstractInferenceSolver, ssvm_wv: WeightVector) {
    var tp: Double = 0
    var fp: Double = 0
    var tn: Double = 0
    var fn: Double = 0
    var total: Double = 0.0

    var i: Int = 0
    while (i < sp.instanceList.size) {
      val gold: ERiStructure = sp.goldStructureList.get(i).asInstanceOf[ERiStructure]
      val prediction:ERiStructure = s_finder.getBestStructure(ssvm_wv, sp.instanceList.get(i)).asInstanceOf[ERiStructure]
      System.out.println(prediction)
      val pred: Boolean = prediction.nodeValues
      val goldval: Boolean = gold.nodeValues
      if (goldval == pred) {
        if (pred == true) {
          tp += 1.0
        }
        else {
          tn += 1.0
        }
      }
      else {
        if (pred ==false) {
          fp += 1.0
        }
        else {
          fn += 1.0
        }
      }
      total += 1.0
      i=i+1

    }
    val precision: Double = tp / (tp + fp)
    val recall: Double = tp / (tp + fn)
    val F1: Double = 2 * precision * recall / (precision + recall)
    println("=========================")
    println("      P        R     F   ")
    // println(String.format("      %-5s %-5s %-5s", "P", "R", "F1"))
    println(precision, recall, F1)//String.format("VALIANT : %4.3f %4.3f %4.3f", precision, recall, F1))
    println("=========================")
    println("Acc = " + (tp + tn) / total)
  }

  def main(args: Array[String]) {
    val modelname: String = "mytest.ssvm.model"
    trainSSVM(modelname)
    System.out.println("\n=== NOW TESTING ===")
    testSequenceSSVM(modelname)
  }
}

File: saul-examples/src/main/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/EntityRelation/SL_SVM/iJLIS/Runner.java
/*
package ml.wolfe.examples.parisa.iJLIS;

import edu.illinois.cs.cogcomp.core.io.LineIO;
import edu.illinois.cs.cogcomp.indsup.inference.AbstractLossSensitiveStructureFinder;
import edu.illinois.cs.cogcomp.indsup.inference.IInstance;
import edu.illinois.cs.cogcomp.indsup.inference.IStructure;
import edu.illinois.cs.cogcomp.indsup.learning.*;
import edu.illinois.cs.cogcomp.indsup.learning.L2Loss.L2LossJLISLearner;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

//import edu.illinois.cs.cogcomp.jlistutorial.SUtils.NodeLabel;

public class Runner {

	*/
/*
	 * Notice that lexm is static.
	 *//*

	static LexManager lexm = new LexManager();

	*/
/**
	 * This modifies the two lists given two it. They are initialized.
	 * 
	 * @param sclist
	 * @param outlist
	 * @param fname
	 * @throws java.io.FileNotFoundException
	 *//*

	public static void readData(List<IInstance> sclist, List<IStructure> outlist, String fname) throws FileNotFoundException {

		ArrayList<String> lines = LineIO.read(fname);

		for(String line : lines){
			List<String> s = Arrays.asList(line.split("\\s+"));
			
			QueryInstance newqi = new QueryInstance(s.subList(0, 3));
			sclist.add(newqi);
			GraphStructure cs = new GraphStructure(newqi, SUtils.binaryToNodeLabel(s.get(3)));
			outlist.add(cs);
		}
	}

	public static String trainSSVM(String modelname) throws Exception {

		// First we read the data.
		// X instance list
		List<IInstance> sclist = new ArrayList();
		// Y instance (structure)
		List<IStructure> outlist = new ArrayList();

		String fname = "/Users/parisakordjamshidi/wolfe-0.1.0/wolfe-examples/src/main/scala/ml/wolfe/examples/parisa/iJLIS/namedata-train";
		readData(sclist, outlist, fname);
		lexm.setAllowNewFeatures(false);

		// Define the problem, and give it data
		StructuredProblem sp = new StructuredProblem();
		sp.input_list = sclist;
		sp.output_list = outlist;

		// Create parameters
		JLISParameters para = new JLISParameters();
		para.total_number_features = lexm.totalNumofFeature();
		para.c_struct = 1;
		para.check_inference_opt = false;

		SimpleInference si = new SimpleInference();
		
		// Actually do the learning
		L2LossJLISLearner learner = new L2LossJLISLearner();
		WeightVector learned_wv = learner.trainStructuredSVM(si, sp, para);
		
		// Package all the parameters together
		SimpleModel model = new SimpleModel();
		model.wv = learned_wv;	
		model.s_finder = new SimpleInference();
		model.lm = lexm;
		
		// This is just a way to serialize the model
		JLISModelIOManager iom = new JLISModelIOManager();
		iom.saveModel(model, modelname);

		return modelname;

	}

	public static void testSequenceSSVM(String model_name) throws Exception {
		// Load the model
		JLISModelIOManager iom = new JLISModelIOManager();
		SimpleModel model = (SimpleModel) iom.loadModel(model_name);
		Runner.lexm = model.lm;
		WeightVector learned_wv = model.wv;
		SimpleInference inference_proc = (SimpleInference) model.s_finder;
		
		// X instance list
		List<IInstance> sclist = new ArrayList();
		// Y instance (structure)
		List<IStructure> outlist = new ArrayList();

		// Load the data
		String fname = "/Users/parisakordjamshidi/wolfe-0.1.0/wolfe-examples/src/main/scala/ml/wolfe/examples/parisa/iJLIS/namedata-trainnamedata-test";
		readData(sclist, outlist, fname);

		// Create a problem with testing data
		StructuredProblem sp = new StructuredProblem();
		sp.input_list = sclist;
		sp.output_list = outlist;

		printTestACC(sp, inference_proc, learned_wv);
	}

	private static void printTestACC(StructuredProblem sp, AbstractLossSensitiveStructureFinder s_finder, WeightVector ssvm_wv) throws IOException,
			Exception {

		double tp = 0;
		double fp = 0;
		double tn = 0;
		double fn = 0;
		double total = 0.0;

		for (int i = 0; i < sp.input_list.size(); i++) {

			GraphStructure gold = (GraphStructure) sp.output_list.get(i);
			GraphStructure prediction = (GraphStructure) s_finder.getBestStructure(ssvm_wv, sp.input_list.get(i));
			System.out.println(prediction);
			
			for (int j = 0; j < gold.nodeValues.size(); j++) {
				SUtils.NodeLabel pred = prediction.nodeValues.get(j);
				SUtils.NodeLabel goldval = gold.nodeValues.get(j);

				// got it right.
				if (goldval == pred) {
					if (pred == SUtils.NodeLabel.VALIANT) {
						tp += 1.0;
					} else {
						tn += 1.0;
					}
				} else { // so I got the answer wrong
					if (pred == SUtils.NodeLabel.VALIANT) {
						fp += 1.0;
					} else {
						fn += 1.0;
					}
				}

				total += 1.0;
			}
		}

		double precision = tp / (tp + fp);
		double recall = tp / (tp + fn);
		double F1 = 2 * precision * recall / (precision + recall);
		System.out.println("=========================");
		System.out.println(String.format("      %-5s %-5s %-5s", "P", "R", "F1"));
//		System.out.println(String.format("COWARDLY: %4.3f %4.3f %4.3f", precision, recall, F1));
		System.out.println(String.format("VALIANT : %4.3f %4.3f %4.3f", precision, recall, F1));
		System.out.println("=========================");
		System.out.println("Acc = " + (tp + tn) / total);
	}

	public static void main(String[] args) throws Exception {
		String modelname = "mytest.ssvm.model";
		trainSSVM(modelname);

		System.out.println("\n=== NOW TESTING ===");
		testSequenceSSVM(modelname);

	}

}
*/
